File: protoxls/ConfigStore.cpp
#include "ConfigStore.h"


ConfigStore::ConfigStore(const Descriptor* descriptor)
{
    stores_ = NULL;
    key_names_ = NULL;
    descriptor_ = descriptor;
}

ConfigStore::~ConfigStore()
{
    descriptor_ = NULL;
    if (stores_ != NULL)
    {
        ConfigStoreMap::iterator it = stores_->begin();
        for (; it != stores_->end(); ++it) {
            delete it->second;
            it->second = NULL;
        }
        delete stores_;
        stores_ = NULL;
    }
}

void ConfigStore::ImportData(Message* data)
{
    datas_.push_back(data);
}

void ConfigStore::ImportData(vector<Message*> datas)
{
    datas_.insert(datas_.end(), datas.begin(), datas.end());
}

vector<string> ConfigStore::GetKeyNames()
{
    return *key_names_;
}

const Descriptor* ConfigStore::GetDescriptor()
{
    return descriptor_;
}

Message* ConfigStore::GetData()
{
    return datas_.front();    
}

ConfigStore* ConfigStore::GetConfig(int num_key)
{
    StoreKey store_key;
    store_key.key_type = KEY_TINTEGER;
    store_key.num_key = num_key;
    return GetConfig(store_key);
}

ConfigStore* ConfigStore::GetConfig(string str_key)
{
    StoreKey store_key;
    store_key.key_type = KEY_TSTRING;
    store_key.str_key = str_key;
    return GetConfig(store_key);
}

ConfigStore* ConfigStore::GetConfig(StoreKey store_key)
{
    ConfigStoreMap::iterator it = stores_->find(store_key);
    if (it == stores_->end()) {
        return NULL;
    }
    return it->second;
}

bool ConfigStore::HasChildren()
{
    return stores_ != NULL;
}

void ConfigStore::ExportKeys(vector<StoreKey>& keys)
{
    ConfigStoreMap::iterator it = stores_->begin();
    for (; it != stores_->end(); ++it)
    {
        keys.push_back(it->first);
    }
}

void ConfigStore::ExportDatas(vector<Message*>& datas)
{
    datas.insert(datas.end(), datas_.begin(), datas_.end());
}

bool ConfigStore::GetKeyVal(const Message& data, string key_name, StoreKey* store_key)
{
    const Reflection* reflection = data.GetReflection();
    const Descriptor* descriptor = data.GetDescriptor();
    const FieldDescriptor* field = descriptor->FindFieldByName(key_name);
    if (field == NULL) {
        proto_error("GetKeyVal key filed not found, key=%s, scheme=%s\n", key_name.c_str(), descriptor->full_name().c_str());
        return false;
    }

    switch (field->cpp_type())
    {
    case FieldDescriptor::CPPTYPE_INT32:
        store_key->key_type = KEY_TINTEGER;
        store_key->num_key = reflection->GetInt32(data, field);
        break;
    case FieldDescriptor::CPPTYPE_UINT32:
        store_key->key_type = KEY_TINTEGER;
        store_key->num_key = reflection->GetUInt32(data, field);
        break;
    case FieldDescriptor::CPPTYPE_INT64:
        store_key->key_type = KEY_TINTEGER;
        store_key->num_key = reflection->GetInt64(data, field);
        break;
    case FieldDescriptor::CPPTYPE_UINT64:
        store_key->key_type = KEY_TINTEGER;
        store_key->num_key = reflection->GetUInt64(data, field);
        break;
    case FieldDescriptor::CPPTYPE_ENUM:
        store_key->key_type = KEY_TINTEGER;
        store_key->num_key = reflection->GetEnumValue(data, field);
        break;
    case FieldDescriptor::CPPTYPE_STRING:
        store_key->key_type = KEY_TSTRING;
        store_key->str_key = reflection->GetString(data, field);
        break;
    default:
        proto_error("GetKeyVal key filed unknow type, key=%s, scheme=%s\n", key_name.c_str(), descriptor->full_name().c_str());
        return false;
    }
    return true;
}

bool ConfigStore::BuildStore(vector<string> key_names)
{
    if (key_names.size() == 0) {
        proto_error("BuildStore keys empty");
        return false;
    }
    
    stores_ = new ConfigStoreMap();
    key_names_ = new vector<string>(key_names);

    string key_name = key_names.front();
    for (size_t i = 0; i < datas_.size(); i++)
    {
        Message* data = datas_[i];
        StoreKey store_key;
        if (!GetKeyVal(*data, key_name, &store_key)) {
            proto_error("BuildStore get keyval fail, key=%s\n", key_name.c_str());
            return false;
        }

        ConfigStoreMap::iterator it = stores_->find(store_key);
        if (it == stores_->end())
        {
            ConfigStore* store = new ConfigStore(descriptor_);
            store->ImportData(data);
            stores_->insert(std::make_pair(store_key, store));
        }
        else
        {
            it->second->ImportData(data);
        }
    }

    if (key_names.size() > 1) {
        vector<string> sub_keys;
        sub_keys.insert(sub_keys.end(), key_names.begin()+1, key_names.end());
        ConfigStoreMap::iterator it = stores_->begin();
        for (; it != stores_->end(); ++it)
        {
            if (!it->second->BuildStore(sub_keys)) {
                proto_error("BuildStore build sub store fail, key=%s\n", key_name.c_str());
                return false;
            }
        }
    }

    return true;
}

File: protoxls/configstore.go
package protoxls

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/jhump/protoreflect/desc"
	"github.com/jhump/protoreflect/dynamic"
)

type StoreKeyType int

const (
	KeyTypeInteger StoreKeyType = iota + 1
	KeyTypeString
)

type StoreKey struct {
	KeyType  StoreKeyType
	NumKey   int64
	StrKey   string
}

func (k StoreKey) String() string {
	if k.KeyType == KeyTypeInteger {
		return fmt.Sprintf("%d", k.NumKey)
	}
	return k.StrKey
}

type ConfigStore struct {
	descriptor *desc.MessageDescriptor
	datas      []*dynamic.Message
	stores     map[StoreKey]*ConfigStore
	keyNames   []string
}

func NewConfigStore(descriptor *desc.MessageDescriptor) *ConfigStore {
	return &ConfigStore{
		descriptor: descriptor,
		datas:      make([]*dynamic.Message, 0),
		stores:     make(map[StoreKey]*ConfigStore),
		keyNames:   make([]string, 0),
	}
}

func (cs *ConfigStore) ImportData(data *dynamic.Message) {
	cs.datas = append(cs.datas, data)
}

func (cs *ConfigStore) ImportDatas(datas []*dynamic.Message) {
	cs.datas = append(cs.datas, datas...)
}

func (cs *ConfigStore) HasChildren() bool {
	return len(cs.stores) > 0
}

func (cs *ConfigStore) BuildStore(keyNames []string) error {
	if len(keyNames) == 0 {
		return nil
	}

	cs.keyNames = keyNames
	keyName := keyNames[0]
	remainingKeys := keyNames[1:]

	for _, data := range cs.datas {
		key, err := cs.getKeyVal(data, keyName)
		if err != nil {
			return fmt.Errorf("failed to get key value: %v", err)
		}

		store, exists := cs.stores[key]
		if !exists {
			store = NewConfigStore(cs.descriptor)
			cs.stores[key] = store
		}

		store.ImportData(data)
	}

	// Recursively build child stores
	for _, store := range cs.stores {
		if err := store.BuildStore(remainingKeys); err != nil {
			return err
		}
	}

	return nil
}

func (cs *ConfigStore) GetData() *dynamic.Message {
	if len(cs.datas) > 0 {
		return cs.datas[0]
	}
	return nil
}

func (cs *ConfigStore) GetConfig(key interface{}) *ConfigStore {
	var storeKey StoreKey
	switch v := key.(type) {
	case int:
		storeKey = StoreKey{KeyType: KeyTypeInteger, NumKey: int64(v)}
	case int32:
		storeKey = StoreKey{KeyType: KeyTypeInteger, NumKey: int64(v)}
	case int64:
		storeKey = StoreKey{KeyType: KeyTypeInteger, NumKey: v}
	case string:
		storeKey = StoreKey{KeyType: KeyTypeString, StrKey: v}
	default:
		return nil
	}

	return cs.stores[storeKey]
}

func (cs *ConfigStore) GetKeyNames() []string {
	return cs.keyNames
}

func (cs *ConfigStore) GetDescriptor() *desc.MessageDescriptor {
	return cs.descriptor
}

func (cs *ConfigStore) ExportKeys() []StoreKey {
	keys := make([]StoreKey, 0, len(cs.stores))
	for key := range cs.stores {
		keys = append(keys, key)
	}
	return keys
}

func (cs *ConfigStore) ExportDatas() []*dynamic.Message {
	return cs.datas
}

func (cs *ConfigStore) getKeyVal(data *dynamic.Message, keyName string) (StoreKey, error) {
	field := cs.descriptor.FindFieldByName(keyName)
	if field == nil {
		return StoreKey{}, fmt.Errorf("field %s not found", keyName)
	}

	value := data.GetField(field)
	if value == nil {
		return StoreKey{}, fmt.Errorf("field %s is nil", keyName)
	}

	switch field.GetType().String() {
	case "TYPE_INT32", "TYPE_SINT32", "TYPE_SFIXED32", "TYPE_UINT32", "TYPE_FIXED32":
		if v, ok := value.(int32); ok {
			return StoreKey{KeyType: KeyTypeInteger, NumKey: int64(v)}, nil
		}
	case "TYPE_INT64", "TYPE_SINT64", "TYPE_SFIXED64", "TYPE_UINT64", "TYPE_FIXED64":
		if v, ok := value.(int64); ok {
			return StoreKey{KeyType: KeyTypeInteger, NumKey: v}, nil
		}
	case "TYPE_STRING":
		if v, ok := value.(string); ok {
			// Try to parse as number first
			if num, err := strconv.ParseInt(v, 10, 64); err == nil {
				return StoreKey{KeyType: KeyTypeInteger, NumKey: num}, nil
			}
			return StoreKey{KeyType: KeyTypeString, StrKey: v}, nil
		}
	}

	return StoreKey{}, fmt.Errorf("unsupported key type for field %s", keyName)
}

// Split utility function
func Split(s, sep string) []string {
	if s == "" {
		return []string{}
	}
	return strings.Split(s, sep)
}